csrc/id_model/id_model.cpp

```diff
@@ -837,4 +845,275 @@ ValGraph IdModel::buildIntersection(
   return intersection;
 }
 
+namespace {
+
+// Check if there's an equivalent expression as iel_expr that uses
+// maybe_promoted_inputs. This is used to avoid redundantly replaying
+// expressions.
+// NOTE: This is currently overly conservative and some
+// opportunities for reuse are lost, althought it doesn't affect
+// the correctness of the analysis.
+Expr* findMatchingExpr(
+    const ExprGroup& iel_expr,
+    const ValGraph& iel_graph,
+    const std::vector<IterDomain*>& maybe_promoted_inputs) {
+  // Grab all eligible uses of the promoted inputs
+  ExprGroups maybe_promoted_input_uses;
+  // Note that any eligible matching expr should be a use of all
+  // inputs in maybe_promoted_input_uses, no matter it's promoted or
+  // not. So it isn't necessary to look at all of
+  // maybe_promoted_input_uses but just need to grab one.
+  for (auto inp_id : maybe_promoted_inputs) {
+    // inp_id may have been just replayed, in which case it should
+    // not exist in the IEL graph. It should be just ignored as it
+    // should not have any use yet.
+    if (!iel_graph.hasGroup(inp_id)) {
+      continue;
```

Comment:
If `inp_id` is just replayed, then shall we just `return nullptr` because it is guaranteed that the `permissiveAreMapped` check below will be false?

Reply:
Yes, you're right. Let me update the code again. Thanks for the suggestion!